### PR TITLE
With one login per subdomain

### PR DESCRIPTION
**Pull request by jscyo**
a.example.com/auth
b.example.com/auth

Related to https://github.com/supertokens/main-website/issues/243

### CODE REVIEW

File: examples/with-one-login-per-subdomain/.env [hunk 1]
```diff
@@ -0,0 +1 @@
+HOST='a.example.com'
```
Comment (NkxxkN) [inline examples/with-one-login-per-subdomain/.env:1]: Is this needed?

File: examples/with-one-login-per-subdomain/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+## Setup
+- Run the command `npm i -d` in the `with-one-login-per-subdomain` directory.
+- Add the following to your hosts configs `/etc/hosts` to allow for multi tenancy locally
+```
+127.0.0.1   a.example.com
+127.0.0.1   b.example.com
+```
+
+> **_NOTE:_**  For the demo, the subdomain for each user is recived from the users email, ex. for `user@a.com`, `a` would be the subdomain. A valid email address is required for password reset and email verification, to use these features in the demo you would need to add that email provider as a subdomain in `/etc/hosts` and add it to the cors domain whitelist on the server.
```
Comment (NkxxkN) [inline examples/with-one-login-per-subdomain/README.md:13]:
```suggestion
> **_NOTE:_**  For the demo, the subdomain is extracted from user's email provider, ex. for `user@a.com`, `a` would be the subdomain. A valid email address is required for password reset and email verification, to use these features in the demo you would need to add that email provider as a subdomain in `/etc/hosts` and add it to the cors domain whitelist on the server.  
```

File: examples/with-one-login-per-subdomain/api-server.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+const cors = require("cors");
+let supertokens = require("supertokens-node");
+let Session = require("supertokens-node/recipe/session");
+let EmailPassword = require("supertokens-node/recipe/emailpassword");
+
+const apiPort = process.env.REACT_APP_API_PORT || 3001;
+const apiDomain = process.env.REACT_APP_API_URL || `http://example.com:${apiPort}`;
+const websitePort = process.env.REACT_APP_WEBSITE_PORT || 3000;
+const websiteDomain = process.env.REACT_APP_WEBSITE_URL || `http://example.com:${websitePort}`
+let whitelist = []
```
Comment (NkxxkN) [inline examples/with-one-login-per-subdomain/api-server.js:11]:
```suggestion
const whitelist = [
  websiteDomain,
  `http://a.example.com:${websitePort}`,
  `http://b.example.com:${websitePort}`
];
```

**Review by NkxxkN — COMMENTED** [review]
- **Can we use a more neutral domain than `example.com`?** 
Forcing anyone who wants to play around with it to overwrite example.com DNS configs is a bit strange. Even though this is the most useless website I agree.
Can we change to `supertokens.dev`? wdyt @jscyo ?

- **Cors**
Is there a way to whitelist all subdomains for CORS? This doesn't scale. Whenever a new company signup, we are asking that they modify their code for CORS. Of course it's only an example but that would be nice to have it automated. 
Accept `*.example.com` ?

Good job on your first PR to the project 💪 !

**Comment by NkxxkN** [comment]
TODO: 
- [ ] Update https://supertokens.io/docs/emailpassword/common-customizations/multi-tenancy/one-login-per-sub-domain with a link to this example.
- Update https://supertokens.io/docs/emailpassword/common-customizations/multi-tenancy/one-login-per-sub-domain code snippets.

File: examples/with-one-login-per-subdomain/.env (hunk 1 above)
Comment (jscyo) [reply examples/with-one-login-per-subdomain/.env:1]: Added that so when the app is started it does not default localhost:3000, i wanted it to default to a.example.com when you start up the app.

**Comment by jscyo** [comment]
> * **Can we use a more neutral domain than `example.com`?**
>   Forcing anyone who wants to play around with it to overwrite example.com DNS configs is a bit strange. Even though this is the most useless website I agree.
>   Can we change to `supertokens.dev`? wdyt @jscyo ?  

Yeah sure, I set it to `example.com` just so it was similar to the docs.

> * **Cors**
>   Is there a way to whitelist all subdomains for CORS? This doesn't scale. Whenever a new company signup, we are asking that they modify their code for CORS. Of course it's only an example but that would be nice to have it automated.
>   Accept `*.example.com` ?  
Sure, we can set it up this.
> 
> Good job on your first PR to the project  !
Thanks!!

**Comment by jscyo** [comment]
### Changes:
* [ ] Change domain `example.com` to more neutral domain `supertokens.dev`
* [ ] Set whitelisted domains to `*.supertokens.dev`

@NkxxkN  These are the requested changes right?

**Comment by NkxxkN** [comment]
Let's keep it consistent with docs then. `example.com` is fine!

File: examples/with-one-login-per-subdomain/api-server.js [hunk 4]
```diff
@@ -8,10 +8,7 @@ const apiPort = process.env.REACT_APP_API_PORT || 3001;
 const apiDomain = process.env.REACT_APP_API_URL || `http://example.com:${apiPort}`;
 const websitePort = process.env.REACT_APP_WEBSITE_PORT || 3000;
 const websiteDomain = process.env.REACT_APP_WEBSITE_URL || `http://example.com:${websitePort}`
-let whitelist = []
-whitelist.push(websiteDomain)
-whitelist.push(`http://a.example.com:${websitePort}`)
-whitelist.push(`http://b.example.com:${websitePort}`)
+let whitelist = /http?:\/\/([a-z0-9]+[.])example[.]com/
```
Comment (NkxxkN) [inline examples/with-one-login-per-subdomain/api-server.js:11]: This will also match `https://attacker.com/http://a.example.com` and would result in bypassing CORS. `/$http?:\/\/([a-z0-9]+[.])example[.]com/` would solve this partially at least.

Comment (jscyo) [reply examples/with-one-login-per-subdomain/api-server.js:11]: Alright, i'll update the whitelist with the above session

File: examples/with-one-login-per-subdomain/.env (hunk 1 above)
Comment (NkxxkN) [reply examples/with-one-login-per-subdomain/.env:1]: Oww cool!

File: examples/with-one-login-per-subdomain/api-server.js (hunk 4 above)
Comment (jscyo) [reply examples/with-one-login-per-subdomain/api-server.js]: Updated the regex, the previous regex we discussed was failing as the origins also have the port, Using the solution mentioned [here](https://stackoverflow.com/questions/14003332/access-control-allow-origin-wildcard-subdomains-ports-and-protocols)

**Review by NkxxkN — APPROVED** [review]

**Comment by rishabhpoddar** [comment]
@NkxxkN if your review comments are taken care of, please resolve them.

File: examples/with-one-login-per-subdomain/src/App.js [hunk 5]
```diff
@@ -1,38 +1,47 @@
...
+    websiteDomain: getAuthDomain()
   },
   recipeList: [
-    EmailPassword.init(),
-    Session.init()
+    EmailPassword.init({
+      emailVerificationFeature: {
+        mode: "REQUIRED"
+      },
+      getRedirectionURL: async (context) => {
```
Comment (rishabhpoddar) [inline examples/with-one-login-per-subdomain/src/App.js:23]: Why is this needed?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]
